# Loops

In this lesson, we’ll be learning about:

* Loops that let us move through each item in a list, called for loops

* Loops that keep going until we tell them to stop, called while loops

* Loops that create new lists, called list comprehensions

In [182]:
dog_breeds = ['french_bulldog', 'dalmatian', 'shihtzu', 'poodle', 'collie']

for breed in dog_breeds:
    print(breed)

french_bulldog
dalmatian
shihtzu
poodle
collie


The general way of writing a for loop is:
<br>
`for <temporary variable> in <list variable>:`
    `<action>`

In [183]:
board_games = ['Settlers of Catan', 'Carcassone', 'Power Grid', 'Agricola', 'Scrabble']

sport_games = ['football', 'football - American', 'hockey', 'baseball', 'cricket']

for game in board_games:
  print(game)

for i in sport_games:
  print(i)

Settlers of Catan
Carcassone
Power Grid
Agricola
Scrabble
football
football - American
hockey
baseball
cricket


***
## Using Range in Loops

So, an easy way to accomplish our "WARNING!" example would be:


In [184]:
for i in range(3):
  print("WARNING!")

WARNING!
WARNING!
WARNING!


In [185]:
promise = "I will not chew gum in class"

for i in range(5):
  print(promise)

I will not chew gum in class
I will not chew gum in class
I will not chew gum in class
I will not chew gum in class
I will not chew gum in class


dd

***
## Using Range in Loops

We’ve iterated through lists that have a discrete beginning and end. However, let’s consider this example:

my_favorite_numbers = [4, 8, 15, 16, 42]
 
`for number in my_favorite_numbers:`

 ` my_favorite_numbers.append(1)`
 
What happens here? __Every time we enter the loop, we add a 1 to the end of the list that we are iterating through. As a result, we never make it to the end of the list! It keeps growing!__

A loop that never terminates is called an infinite loop. These are very dangerous for your code!

__A program that hits an infinite loop often becomes completely unusable. The best course of action is to never write an infinite loop.__

Note: If you accidentally stumble into an infinite loop while developing on your own machine, __you can end the loop by using control + c to terminate the program.__ If you’re writing code in our online editor, you’ll need to refresh the page to get out of an infinite loop!

***
## Continue

We often want to use a for loop to search through a list for some value:

`items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]`
 
`#we want to check if the item with ID "knit_dress" is on sale:`

`for item in items_on_sale:`

`if item == "knit_dress":`
 
`print("Knit Dress is on sale!")`
    
This code goes through each item in items_on_sale and checks for a match. After we find that "knit_dress" is in the list items_on_sale, we don’t need to go through the rest of the items_on_sale list. Since it’s only 5 elements long, iterating through the entire list is not a big deal in this case. But what if items_on_sale had 1000 items after "knit_dress"? What if it had 100,000 items after "knit_dress"?

You can stop a for loop from inside the loop by using break. When the program hits a break statement, control returns to the code outside of the for loop. For example:

`items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]`
 
`print("Checking the sale list!")`

`for item in items_on_sale:`

`  print(item)`

  `if item == "knit_dress":`
  
   ` break`
   
`print("End of search!")`

This would produce the output:

* `Checking the sale list!`
* `Blue_shirt`
* `Striped_socks`
* `knit_dress`
* `End of search!`

We didn’t need to check "red_headband" or "dinosaur_onesie" at all!

***
## Break

We often want to use a for loop to search through a list for some value:

`items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]`
 
`#we want to check if the item with ID "knit_dress" is on sale:`

`for item in items_on_sale:`

`if item == "knit_dress":`
 
`print("Knit Dress is on sale!")`
    
This code goes through each item in items_on_sale and checks for a match. After we find that "knit_dress" is in the list items_on_sale, we don’t need to go through the rest of the items_on_sale list. Since it’s only 5 elements long, iterating through the entire list is not a big deal in this case. But what if items_on_sale had 1000 items after "knit_dress"? What if it had 100,000 items after "knit_dress"?

You can stop a for loop from inside the loop by using break. When the program hits a break statement, control returns to the code outside of the for loop. For example:

`items_on_sale = ["blue_shirt", "striped_socks", "knit_dress", "red_headband", "dinosaur_onesie"]`
 
`print("Checking the sale list!")`

`for item in items_on_sale:`

`  print(item)`

  `if item == "knit_dress":`
  
   ` break`
   
`print("End of search!")`

This would produce the output:

* `Checking the sale list!`
* `Blue_shirt`
* `Striped_socks`
* `knit_dress`
* `End of search!`

We didn’t need to check "red_headband" or "dinosaur_onesie" at all!

***
## Continue

When we’re iterating through lists, we may want to skip some values. Let’s say we want to print out all of the numbers in a list, unless they’re negative. We can use continue to move to the next i in the list:

`big_number_list = [1, 2, -1, 4, -5, 5, 2, -9]`
 
`for i in big_number_list:`

 ` if i < 0:`
 
  `  continue`
  
 `print(i)`
 
This would produce the output:

* 1
* 2
* 4
* 5
* 2

Every time there was a negative number, the continue keyword moved the index to the next value in the list, without executing the code in the rest of the for loop.

In [186]:
ages = [12, 38, 34, 26, 21, 19, 67, 41, 17]

for i in ages:
  if i < 21:
    continue #si la edad edad es mayor que 21, no llega al print,sino que vuelve a iniciar otro ciclo
  print(i)

38
34
26
21
67
41


***
## While


__While Loops__

The while loop performs a set of code until some condition is reached.

While loops can be used to iterate through lists, just like for loops:

`dog_breeds = ['bulldog', 'dalmation', 'shihtzu', 'poodle', 'collie']`

`index = 0`

`while index < len(dog_breeds):`

 ` print(dog_breeds[index])`
  
  `index += 1`
  
Every time the condition of the while loop (in this case, index < len(dog_breeds)) is satisfied, the code inside the while loop runs.

While loops can be useful when you don’t know how many iterations it will take to satisfy a condition.

In [187]:
dog_breeds = ['bulldog', 'dalmation', 'shihtzu', 'poodle', 'collie']
 
index = 0
while index < len(dog_breeds):
  print(dog_breeds[index])
  index += 1


bulldog
dalmation
shihtzu
poodle
collie


In [188]:
all_students = ["Alex", "Briana", "Cheri", "Daniele", "Dora", "Minerva", "Alexa", "Obie", "Arius", "Loki"]
students_in_poetry = []

while len(students_in_poetry) < 6:
  student = all_students.pop()
  students_in_poetry.append(student)
  
print(students_in_poetry)

['Loki', 'Arius', 'Obie', 'Alexa', 'Minerva', 'Dora']


***
## Nested Loops

__Nested Loops__

We have seen how we can go through the elements of a list. What if we have a list made up of multiple lists? How can we loop through all of the individual elements?

Suppose we are in charge of a science class, that is split into three project teams:

`project_teams = [["Ava", "Samantha", "James"], ["Lucille", "Zed"], ["Edgar", "Gabriel"]]`

If we want to go through each student, we have to put one loop inside another:

`for team in project_teams:`

  `for student in team:`
  
   ` print(student)`
   
This results in:

* Ava
* Samantha
* James
* Lucille
* Zed
* Edgar
* Gabriel

In [189]:
project_teams = [["Ava", "Samantha", "James"], ["Lucille", "Zed"], ["Edgar", "Gabriel"]]

for team in project_teams:

  for student in team:
  
   print(student)
   

Ava
Samantha
James
Lucille
Zed
Edgar
Gabriel


In [190]:
sales_data = [[12, 17, 22], [2, 10, 3], [5, 12, 13]]

scoops_sold= 0

for location in sales_data:
 for element in location:
   scoops_sold += element

print(scoops_sold)

96



## List Comprehensions
Let’s say we have scraped a certain website and gotten these words:

`words = ["@coolguy35", "#nofilter", "@kewldawg54", "reply", "timestamp", "@matchamom", "follow", "#updog"]`

We want to make a new list, called usernames, that has all of the strings in words with an '@' as the first character. We know we can do this with a for loop:

`words = ["@coolguy35", "#nofilter", "@kewldawg54", "reply", "timestamp", "@matchamom", "follow", "#updog"]`
`usernames = []`
 
`for word in words:`

 ` if word[0] == '@':`
 
   ` usernames.append(word)`
   
First, we created a new empty list, usernames, and as we looped through the words list, we added every word that matched our criterion. Now, the usernames list looks like this:

`>>> print(usernames)`

`["@coolguy35", "@kewldawg54", "@matchamom"]`

Python has a convenient shorthand to create lists like this with one line:

`usernames = [word for word in words if word[0] == '@']`

This is called a list comprehension. It will produce the same output as the for loop did:

`["@coolguy35", "@kewldawg54", "@matchamom"]`

This list comprehension:

* Takes an element in words
* Assigns that element to a variable called word
* Checks if word[0] == '@', and if so, it adds word to the new list, usernames. If not, nothing happens.
* Repeats steps 1-3 for all of the strings in words

Note: if we hadn’t done any checking (let’s say we had omitted if word[0] == '@'), the new list would be just a copy of words:

`usernames = [word for word in words]`

`#usernames is now ["@coolguy35", "#nofilter", "@kewldawg54", "reply", "timestamp", "@matchamom", "follow", "#updog"]`

## More List Comprehensions

Let’s say we’re working with the usernames list from the last exercise:

`print(usernames)`
`["@coolguy35", "@kewldawg54", "@matchamom"]`

We want to create a new list with the string " please follow me!" added to the end of each username. We want to call this new list messages. We can use a list comprehension to make this list with one line:




In [191]:
usernames = ["@coolguy35", "@kewldawg54", "@matchamom"]

messages = [user + " please follow me!" for user in usernames]

print(messages)

['@coolguy35 please follow me!', '@kewldawg54 please follow me!', '@matchamom please follow me!']


What it does: 

* Takes a string in usernames
* Assigns that string to a variable called user
* Adds “ please follow me!” to user
* Appends that concatenation to the new list called messages
* Repeats steps 1-4 for all of the strings in usernames

Now, messages contains these values:

`["@coolguy35 please follow me!", "@kewldawg54 please follow me!", "@matchamom please follow me!"]`

Being able to create lists with modified values is especially useful when working with numbers. Let’s say we have this list:



In [192]:
my_upvotes = [192, 34, 22, 175, 75, 101, 97] 

We want to add 100 to each value. We can accomplish this goal in one line:

In [193]:
updated_upvotes = [vote_value + 100 for vote_value in my_upvotes]
print(updated_upvotes)


[292, 134, 122, 275, 175, 201, 197]


This list comprehension:

* Takes a number in my_upvotes
* Assigns that number to a variable called vote_value
* Adds 100 to vote_value
* Appends that sum to the new list updated_upvotes
* Repeats steps 1-4 for all of the numbers in my_upvotes

Now, updated_upvotes contains these values:

`[292, 134, 122, 275, 175, 201, 197]`

In [194]:
celsius = [0, 10, 15, 32, -5, 27, 3]

fahrenheit = [(temperature * 9/5 +32) for temperature in celsius] #uses the formula to convert celsius in farenheit

print(fahrenheit)

[32.0, 50.0, 59.0, 89.6, 23.0, 80.6, 37.4]


In [8]:
![title]("https://content.codecademy.com/courses/matplotlib/rainfall.png")

/bin/bash: -c: line 0: syntax error near unexpected token `"https://content.codecademy.com/courses/matplotlib/rainfall.png"'
/bin/bash: -c: line 0: `[title]("https://content.codecademy.com/courses/matplotlib/rainfall.png")'


***
## Final excercise: Lesson Review

In [195]:
single_digits = range(0,10)
print(list(single_digits))
squares = []
cubes = [(cubedNumber ** 3) for cubedNumber in single_digits]

for digit in single_digits:
  squares.append((digit ** 2))
  
print(squares)
print(cubes)


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 8, 27, 64, 125, 216, 343, 512, 729]


***
## Looping Techniques

* When looping through dictionaries, the key and corresponding value can be retrieved at the same time using the items() method.

In [196]:
knights = {'gallahad': 'the pure', 'robin': 'the brave'}
for k, v in knights.items():
    print(k, v)

gallahad the pure
robin the brave


* When looping through a sequence, the position index and corresponding value can be retrieved at the same time using the enumerate() function.

In [197]:
for i, v in enumerate(['tic', 'tac', 'toe']):
    print(i, v)




0 tic
1 tac
2 toe


* To loop over two or more sequences at the same time, the entries can be paired with the zip() function.

In [198]:
questions = ['name', 'quest', 'favorite color']
answers = ['lancelot', 'the holy grail', 'blue']
for q, a in zip(questions, answers):
    print('What is your {0}?  It is {1}.'.format(q, a))


What is your name?  It is lancelot.
What is your quest?  It is the holy grail.
What is your favorite color?  It is blue.


* To loop over a sequence in reverse, first specify the sequence in a forward direction and then call the reversed() function.

In [199]:
for i in reversed(range(1, 10, 2)):
  print(i)

9
7
5
3
1


* To loop over a sequence in sorted order, use the sorted() function which returns a new sorted list while leaving the source unaltered.


In [200]:
basket = ['apple', 'orange', 'apple', 'pear', 'orange', 'banana']
for i in sorted(basket):
  print(i)

apple
apple
banana
orange
orange
pear


* Using set() on a sequence eliminates duplicate elements. The use of sorted() in combination with set() over a sequence is an idiomatic way to loop over unique elements of the sequence in sorted order.

In [201]:
basket = ['apple', 'orange', 'apple', 'pear', 'orange', 'banana']
for f in sorted(set(basket)):
    print(f)

apple
banana
orange
pear


* It is sometimes tempting to change a list while you are looping over it; however, it is often simpler and safer to create a new list instead.

In [202]:
>>> import math
>>> raw_data = [56.2, float('NaN'), 51.7, 55.3, 52.5, float('NaN'), 47.8]
>>> filtered_data = []
>>> for value in raw_data:
...     if not math.isnan(value):
...         filtered_data.append(value)
...
>>> filtered_data

[56.2, 51.7, 55.3, 52.5, 47.8]